Add PosgreSQL 9.2 image

mnagy opened the pull request:
It is based on the MySQL image and works similarly. You must set
`POSTGRESQL_{USER,PASSWORD,DATABASE}` environment variables and can
optionally set the 'postgres' db user password using the
`POSTGRESQL_ADMIN_PASSWORD` variable.

There are few minor outstanding issues that are marked and commented
with a `TODO`. Ideas and suggestions would be appreciated.

You can test the image with `make test && make test TARGET=rhel7`.

Documentation in README is still missing, but wanted early feedback.

mnagy commented on 9.2/run-postgresql.sh:
```diff
@@ -0,0 +1,83 @@
+
+if [ ! -f "$PGDATA/postgresql.conf" ]; then
+
+	chown postgres.postgres "$PGDATA"
```
A note about USER. We run this as root, but only thing that really requires root privileges is this chown. I would love to run the whole entrypoint as user postgres. However, we would have to make sure the volume is owned by the postgres user (uid 26).

Any ideas if this is possible?

mfojtik replied:
Why not chown in Dockerfile then?

soltysh replied:
Nope if you have mounted volume `$PGDATA`.

mfojtik commented on 9.2/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+ENV PGDATA /var/lib/pgsql/data
+# TODO: We should consider if this should be changeable. Can affect performance.
+#       http://www.postgresql.org/docs/9.2/static/locale.html
+ENV LANG en_US.utf8
```
Since we are not doing this for any other image, I would vote for nuking this line. You can set it in run-postgresql.sh (and users can change it by using -e)

soltysh replied:
+1

mfojtik commented on 9.2/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+#       http://www.postgresql.org/docs/9.2/static/locale.html
+ENV LANG en_US.utf8
+
+MAINTAINER  Martin Nagy <mnagy@redhat.com>
```
Move this after FROM

mfojtik commented on 9.2/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+
+COPY run-postgresql.sh /usr/local/bin/
+
+VOLUME ["/var/lib/pgsql/data"]
```
why not have /var/lib/pgsql as VOLUME? Then you can persistently store your config files

mnagy replied:
/var/lib/pgsql/data is a commonly used path for a database cluster, which also contains within the configuration that is necessary, so config files are persistently stored. It didn't make much sense to me to use /var/lib/pgsql. Besides, I think that while experimenting I did see some psql history files there, or something like that. Probably not relevant to us, but stil..

mfojtik commented on 9.2/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+
+VOLUME ["/var/lib/pgsql/data"]
+
+# TODO: Can we run as non-root?
```
that is requirement :-)

soltysh commented on 9.2/test/run:
```diff
@@ -0,0 +1,105 @@
+
+IMAGE_NAME=${IMAGE_NAME-openshift/postgresql-92-centos7-candidate}
+
+CIDFILE_DIR=$(mktemp --suffix=mysql_test_cidfiles -d)
```
mysql :wink: ?

mnagy replied:
Haha, good catch, I thought I killed all mysql references :)

mfojtik commented on 9.2/run-postgresql.sh:
```diff
@@ -0,0 +1,83 @@
+valid_psql_password   "$POSTGRESQL_PASSWORD" || usage
+valid_psql_identifier "$POSTGRESQL_DATABASE" || usage
+
+# Make sure env variables don't propagate to mysqld process.
```
copy&paste, right?

mnagy replied:
Yep, will fix.

jhadvig commented on 9.2/run-postgresql.sh:
```diff
@@ -0,0 +1,83 @@
+	echo "You must specify following environment variables:"
+	echo "  \$POSTGRESQL_USER (regex: '$psql_identifier_regex')"
+	echo "  \$POSTGRESQL_PASSWORD (regex: '$psql_password_regex')"
+	echo "  \$POSTGRESQL_DATABASE (regex: '$psql_identifier_regex')"
```
In MongoDB I am creating the `production` database if none is defined. I think some similar approach we had in v2. Wouldn't it be better to have it in all DB images (also MySQL) ?

soltysh replied:
I'd rather see a failure when that name isn't specified. It's required the
same way username and password is. This is the behavior I'd vote for all
images.

On 6:27pm, Thu, Mar 19, 2015 Jakub Hadvig <notifications@github.com> wrote:

> In 9.2/run-postgresql.sh
> <https://github.com/openshift/postgresql/pull/1#discussion_r26777138>:
>
> > @@ -0,0 +1,83 @@
> > +#!/bin/bash -e
> > +
> > +# Be paranoid and stricter than we should be.
> > +psql_identifier_regex='^[a-zA-Z_][a-zA-Z0-9_]*$'
> > +psql_password_regex='^[a-zA-Z0-9_~!@#$%^&*()-=<>,.?;:|]+$'
> > +
> > +function usage {
> > +	echo "You must specify following environment variables:"
> > +	echo "  \$POSTGRESQL_USER (regex: '$psql_identifier_regex')"
> > +	echo "  \$POSTGRESQL_PASSWORD (regex: '$psql_password_regex')"
> > +	echo "  \$POSTGRESQL_DATABASE (regex: '$psql_identifier_regex')"
>
> In MongoDB I am creating the production database if none is defined. I
> think some similar approach we had in v2. Wouldn't it be better to have it
> in all DB images (also MySQL) ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/postgresql/pull/1/files#r26777138>.
>

mnagy replied:
I think we discussed somewhere for MySQL image that we don't want optional vars and that we recommend that in the image writing guide, that's why it's not optional.

Personally, I'd be in favour of that, it would make using the image outside Openshift a LOT easier.

jhadvig replied:
I dont recall any discussion about optional vars. Also the MYSQL_ROOT_PASSWORD var is optional, is it not ?
I would go with similar behavior as the MongoDB, less typing => happy users :)

mnagy replied:
Hm. We can talk about that on scrum I suppose?

mfojtik replied:
database name is required, always... we need to fix mongo

mnagy replied:
@jhadvig also, MYSQL_ROOT_PASSWORD is a special case since it doesn't have a default but not specifying it alters the behavior (remote login to root account is disabled).

mfojtik replied:
if you docker exec to running container and run psql, does it connect to postres? it should without having to specify any arguments... please test that

mnagy replied:
You'll get command not found error because of SCL :) But yeah, if you run it through scl, psql will work as is.

jhadvig commented on 9.2/test/run:
```diff
@@ -0,0 +1,105 @@
+	done
+	rmdir $CIDFILE_DIR
+}
+trap cleanup EXIT
```
+ `SIGINT`

mnagy replied:
Good catch.

mnagy replied:
I take that back; it now calls cleanup two times. EXIT trap is invoked after an interrupt as well.

jhadvig commented on 9.2/test/run:
```diff
@@ -0,0 +1,105 @@
+# Tests.
+
+USER=user PASS=pass run_tests no_admin
+USER=user1 PASS=pass1 ADMIN_PASS=r00t run_tests admin
```
You should be also testing the admin user, not only standalone user.

mnagy replied:
I'll add in access tests that we do in mysql.

mfojtik replied:
also test failures ... no user set, etc...

mnagy replied:
You mean no env vars set when creating the container? Please elaborate.

mfojtik replied:
invalid values, empty values.... exercise the validate functions in image

mnagy commented:
Addressed all issues plus something extra :) PTAL

mfojtik commented on 9.2/run-postgresql.sh:
```diff
@@ -0,0 +1,81 @@
+
+# Enable utf8 support. This might affect performance:
+# http://www.postgresql.org/docs/9.2/static/locale.html
+export LANG=en_US.utf8
```
this needs to be tuneable via env var... why is default not enough? can't you set this in postgres config file instead?

mnagy replied:
Right, will fix that tomorrow so LANG can be overridden. I might, but setting the LANG var is the simplest way to do it, I didn't see any reason to modify config file because of it. Btw, it mostly affects initdb. And yeah, the default is probably faster a bit, but without utf support, at least that's how I understand it.

mfojtik commented:
LGTM. Nice job @mnagy, kudos!

mnagy commented:
Some enhancements:
No further need to use scl when doing docker exec, added contrib directory. Added tests for env variable validations, simplified that code a bit as well. PTAL.

mnagy commented:
Very small change added. We're now also validating maximum user/db name length. Also added tests for that.

mnagy commented:
@jhadvig @soltysh can I get more reviews in addition to @mfojtik's, please?

jhadvig commented on 9.2/run-postgresql.sh:
```diff
@@ -0,0 +1,91 @@
+psql_identifier_regex='^[a-zA-Z_][a-zA-Z0-9_]*$'
+psql_password_regex='^[a-zA-Z0-9_~!@#$%^&*()-=<>,.?;:|]+$'
+
+function usage {
```
could you maybe do the function declaration in this file more consistent with the test/run script?

mnagy replied:
Done. PTAL

soltysh commented on 9.2/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+    test "$(id postgres)" = "uid=26(postgres) gid=26(postgres) groups=26(postgres)"
+
+COPY run-postgresql.sh /usr/local/bin/
+COPY contrib/.bashrc /var/lib/pgsql/
```
Can you do that also for https://github.com/openshift/mongodb/ and https://github.com/openshift/mysql/, please :smile_cat:

soltysh replied:
Even more, can you add that to all our images :sweat_smile:

mnagy replied:
@soltysh For mysql, see https://github.com/openshift/mysql/pull/24
For MongoDB, @jhadvig will do that. :)

soltysh replied:
many kudos @mnagy

mnagy replied:
@soltysh thanks :) I'd also appreciate a certain four-letter acronym :P

jhadvig commented on 9.2/test/run:
```diff
@@ -0,0 +1,169 @@
+}
+
+# Make sure the invocation of docker run fails.
+function assert_image_creation_fails {
```
It should be `assert_container_creation_fails`

mnagy commented:
Added SCL info into README.md, few minor fixes. PTAL

soltysh commented:
Here's mine LGTM :)

jhadvig commented:
@mnagy LGTM, good job :)
Merging

rhcarvalho commented on 9.2/test/run:
```diff
@@ -0,0 +1,169 @@
+}
+
+function postgresql_cmd() {
+	docker run --rm -e PGPASSWORD="${PASS}" $IMAGE_NAME psql postgresql://$USER@$CONTAINER_IP:5432/db "$@"
```
This uses `PASS` and `USER` as global variables, so we're never actually testing the access of the admin user...

soltysh replied:
Can you open a bug report against this repo, so it doesn't get forgotten?

rhcarvalho replied:
I was wrong. In fact, what happens is that `assert_login_access` shadows `USER` and `PASS` with local variables of the same name, and because in bash [local variables are visible in the scope of a function *and its children*](http://tldp.org/LDP/abs/html/localvar.html#FTN.AEN18568), `postgresql_cmd` eventually uses those local variables when called from within `assert_login_access`.

Needless to say, that's very very error prone and hard to reason about...
